File: android/app/src/main/java/network/mysterium/vpn/connection/ConnectionChecker.kt
/*
 * Copyright (C) 2019 The "mysteriumnetwork/mysterium-vpn-mobile" Authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package network.mysterium.vpn.connection

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.util.Log
import java.lang.IllegalStateException

/**
 * Starts ConnectionCheckerService periodically at a given interval.
 */
class ConnectionChecker(private val context: Context, private val interval: Long) {
  private var running: Boolean = false

  fun start() {
    if (running) {
      return
    }

    Log.d(TAG, "Starting ConnectionChecker")
    running = true
    this.loopService(ignoreLastStatus = true)
  }

  fun stop() {
    if (!running) {
      return
    }

    Log.d(TAG, "Stopping ConnectionChecker")
    running = false
  }

  private fun loopService(ignoreLastStatus: Boolean) {
    if (!running) {
      return
    }
    startService(ignoreLastStatus)

    Handler().postDelayed({ this.loopService(ignoreLastStatus = false) }, interval)
  }

  private fun startService(ignoreLastStatus: Boolean) {
    Log.d(TAG, "Starting ConnectionCheckerService")
    val service = Intent(context, ConnectionCheckerService::class.java)
    val bundle = Bundle()
    bundle.putBoolean("ignoreLastStatus", ignoreLastStatus)
    service.putExtras(bundle)
    try {
      context.startService(service)
    } catch (e: IllegalStateException) {
      // We stop checker, because app is in a state where the service can not be started.
      // This is usually because app was in the background for too long.
      stop()
    }
  }

  companion object {
    private const val TAG = "ConnectionChecker"
  }
}

File: src/app/domain/connection-checker.ts
/*
 * Copyright (C) 2019 The "mysteriumnetwork/mysterium-vpn-mobile" Authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { ConnectionStatus } from 'mysterium-tequilapi/lib/dto/connection-status'
import ConnectionAdapter from '../adapters/connection/connection-adapter'
import NotificationAdapter from '../adapters/notification/notification-adapter'
import translations from '../translations'

type ConnectionCheckerData = {
  ignoreLastStatus?: boolean
}

class ConnectionChecker {
  private lastStatus: ConnectionStatus | null = null
  private running: boolean = false
  private readonly NAME = 'ConnectionChecker'

  constructor (private connectionAdapter: ConnectionAdapter, private notificationAdapter: NotificationAdapter) {}

  public async run (data: ConnectionCheckerData) {
    if (this.running) {
      return
    }

    this.running = true

    if (data.ignoreLastStatus) {
      this.lastStatus = null
    }

    const status = await this.fetchStatus()
    console.log(`${this.NAME}, status fetched:`, status)
    if (this.wasDisconnected(this.lastStatus, status)) {
      this.showDisconnectedNotification()
    }
    this.lastStatus = status

    this.running = false
  }

  private async fetchStatus (): Promise<ConnectionStatus | null> {
    try {
      const statusDTO = await this.connectionAdapter.fetchStatus()
      return statusDTO.status
    } catch (err) {
      console.log(`${this.NAME}, fetching connection status failed:`, err)
      return null
    }
  }

  private wasDisconnected (oldStatus: ConnectionStatus | null, newStatus: ConnectionStatus | null): boolean {
    return oldStatus === 'Connected' && newStatus !== this.lastStatus
  }

  private showDisconnectedNotification () {
    const title = translations.DISCONNECTED_NOTIFICATION.TITLE
    const message = translations.DISCONNECTED_NOTIFICATION.MESSAGE
    this.notificationAdapter.show(title, message)
  }
}

export { ConnectionChecker }
